domain/include/cstone/halos/exchange_halos_gpu.cuh

```diff
@@ -109,6 +117,10 @@ void haloExchangeGpu(int epoch,
         mpiSendGpuDirect(sendPtr, sendBytes, destinationRank, haloExchangeTag, sendRequests, sendBuffers);
         sendPtr += sendBytes;
     }
+    tlast      = Clock::now();
+    timings[0] = std::chrono::duration_cast<Time>(tlast - tstart).count();
+    // std::cout << "mpiSendAsync = " << timings[0] << std::endl;
+    tstart = tlast;
```

Comment:
This is an asynchronous function call, the timings that you record here will be wrong; this is not measuring the time it took to send.

Reply:
True, the second value you're returning could be different from the total as it might be necessary to wait for the sends to be completed. But the first value measures the time it takes for buffer setup and the asynchronous MPI_Isend calls to return, which is not well defined as it's not clear how much overlap there is between communication and preparing the buffers between the first to time-stamps.

It doesn't hurt to have the propagator profiling calls in the main production branch, as long as they replace the existing `timer.step` calls that only measure `rank 0` (instead of complementing them).